[clang-tidy] Add user-defined functions to `bugprone-unsafe-functions` check

Discookie opened the pull request:
Adds the check option `bugprone-unsafe-functions.CustomFunctions` to be able to match user-defined functions as part of the checker.
Adds the option `bugprone-unsafe-functions.ReportDefaultFunctions` to disable reporting the default set of functions as well.

The functions names are matched using the same mechanism as the `matchesAnyListedName` tidy matcher, documented in `unsafe-functions.rst`.

Discookie committed:
[clang-tidy] Add user-defined functions to bugprone-unsafe-functions check

Discookie committed:
[clang-tidy] Make MatchesAnyListedNameMatcher::NameMatcher public

The same matching mechanism should be used both there and within bugprone-unsafe-functions check.

Discookie committed:
[clang-tidy] Add regex support and documentation for C++ functions for bugprone-unsafe-functions check

whisperity commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -7,6 +7,8 @@
 //===----------------------------------------------------------------------===//
 
 #include "UnsafeFunctionsCheck.h"
+#include "../utils/Matchers.h"
```
(This include is superfluous, ClangTidyCheck automatically pulls the related header in. Yes, I'm always burnt by clangd with this as well. :wink:)

Discookie replied:
Not in this case! utils/Matchers.h is not included in ClangTidyCheck (but it is included in UnsafeFunctionsCheck.h, so... whoops)

whisperity commented on clang-tools-extra/clang-tidy/utils/Matchers.h:
```diff
@@ -136,6 +128,15 @@ class MatchesAnyListedNameMatcher
     }
   };
 
+  bool matches(
+      const NamedDecl &Node, ast_matchers::internal::ASTMatchFinder *Finder,
+      ast_matchers::internal::BoundNodesTreeBuilder *Builder) const override {
+    return llvm::any_of(NameMatchers, [&Node](const NameMatcher &NM) {
+      return NM.match(Node);
+    });
+  }
+
+private:
```
Why was this change needed? The other changes in the code do not indicate this function getting called anywhere.

Discookie replied:
The diff here highlights the wrong part of the change:
https://github.com/llvm/llvm-project/pull/106350/commits/47c1a5694fed5769bb6bbc437e31276efe41d520

I made the contained NameMatcher class public, but the diff considered the other function to be moved.

whisperity commented:
High-level design question: Why is `CustomAnnexKFunctions` a thing?

Discookie commented:
`CustomAnnexKFunctions` exists because the checker has two different internal matching modes for whether AnnexK is enabled or not. This both depends on `__STDC_LIB_EXT1__` (defined by system) `__STDC_WANT_LIB_EXT1__` (defined by the user), so it's reasonable to expect that a project could have differing availability of AnnexK.

I could rename CustomNormalFunctions to CustomFunctions or something to make it clearer that it's always available.

Discookie committed:
Merge branch 'main' into unsafe-expand

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -74,6 +77,40 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
+*(Note: The regular expressions do not support negative* ``(?!)`` *matches)*
+
+The `reason` is optional and is used to provide additional information about the
+reasoning behind the replacement. The default reason is ``is marked as unsafe``.
```
```suggestion
reasoning behind the replacement. The default reason is `is marked as unsafe`.
```

Option values should be enclosed in single back-ticks. Same below.

EugeneZelenko reviewed (commented):
Please mention changes in Release Notes.

5chmidti commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -86,6 +123,22 @@ Options
+    When `true`, the check reports the default set of functions.
+    Default is `true`.
+
+.. option:: CustomNormalFunctions
+
+    A comma-separated list of regular expressions, their replacements, and an
+    optional reason. For more information, see `Custom functions
+    <unsafe-functions.html#custom-functions>`_.
+
+.. option:: CustomAnnexKFunctions
+
+    A comma-separated list of regular expressions, their replacements, and an
+    optional reason. For more information, see `Custom functions
+    <unsafe-functions.html#custom-functions>`_.
```
What do you think about:

> A semicolon-separated list of specifications, each containing a regular expression, a replacement, and an optional reason, separated by a comma.

? I think there is still room for improvement, but I'm worried about the not mentioned semicolon/structure (even though you link to the documentation).

Discookie replied:
> A semicolon-separated list of custom functions to be matched. A matched
    function contains a regular expression, the name of the replacement
    function, and an optional reason, separated by comma.

Should be good enough for these purposes.

5chmidti reviewed (commented):
Just some small things

Discookie committed:
Review fixes

Do not add invalid options to the functions list
Rewrite description for options
Minor formatting fixes

Discookie commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -127,57 +139,157 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
+          "invalid configuration value '%0' for option '%1'; "
+          "expected a replacement function name")
+          << Name.trim() << OptionName;
+    }
```
For some reason I thought this call was treated as a fatal error, fixed.

Discookie commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -186,16 +298,62 @@ void UnsafeFunctionsCheck::check(const MatchFinder::MatchResult &Result) {
+      }
+    }
+
+    assert(false && "No custom function was matched.");
```
Only this one, the others have proper invariants attached.

Discookie committed:
Remove redundant import

Discookie commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -19,6 +19,9 @@ The check implements the following rules from the CERT C Coding Standard:
 
+The following functions are reported if `ReportDefaultFunctions
+<unsafe-functions.html#cmdoption-arg-ReportDefaultFunctions>`_ is enabled.
+
```
I hope I did it correctly, I'm not that familiar with RST syntax.

whisperity replied:
Referring to the `.html` directly is definitely not good because rst can be lowered to different markup syntaxes.

You need to create (or check if `opt::` automatically does?) a referable entity, and then `:ref:` can be used to make the link clickable. So you will use the ID or "label" you allocated in the document instead of relying on the automatically generated, HTML-specific anchor.

Discookie commented:
Internal was the wrong word there, the distinction is only about whether AnnexK is available or not, in the context of the checked code. We certainly could eliminate `CustomAnnexKFunctions`, but that would mean that it's the user's responsibility to check whether AnnexK is available, outside of the code being checked.
Ultimately I don't think it would get too much use, if for nothing other than being able to reproduce the default-checked functions with just the `Custom__Functions`.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -127,57 +138,159 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
+static std::vector<UnsafeFunctionsCheck::CheckedFunction>
+parseCheckedFunctions(StringRef Option, StringRef OptionName,
+                      ClangTidyContext *Context) {
+  std::vector<StringRef> Functions = utils::options::parseStringList(Option);
```
```suggestion
  const std::vector<StringRef> Functions = utils::options::parseStringList(Option);
```

Discookie committed:
Remove CustomAnnexKFunctions and related logic

Ultimately it would not get much use outside of being able to reproduce
the default check logic.

Discookie committed:
Remove CustomAnnexKFunctions from docs

Discookie committed:
Attempt to fix rst references (again)

Discookie committed:
Small constness fix

Discookie commented:
Removed the normal/AnnexK distinction as I think it would get little use as it's a rare edge case, and it made everything more confusing.

Szelethus reviewed (commented):
Alright, I think this patch is good from a design standpoint! I'll get into the nitty-gritty in a bit.

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -127,57 +134,137 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
+  Result.reserve(Functions.size());
+
+  for (StringRef Function : Functions) {
+    if (Function.empty()) {
+      continue;
+    }
```
nit: single statement with braces

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/UnsafeFunctionsCheck.cpp:
```diff
@@ -127,57 +134,137 @@ static bool isAnnexKAvailable(std::optional<bool> &CacheVar, Preprocessor *PP,
+      continue;
+    }
+
+    auto [Name, Rest] = Function.split(',');
+    auto [Replacement, Reason] = Rest.split(',');
```
nit: Please declare both of these `const`

5chmidti reviewed (approved):
This looks good from my side.

Discookie committed:
Add support for no specified replacement function

Discookie committed:
Review fixes

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -45,8 +47,7 @@ The following functions are always checked, regardless of *Annex K* availability
 
-If `ReportMoreUnsafeFunctions
-<unsafe-functions.html#cmdoption-arg-ReportMoreUnsafeFunctions>`_ is enabled,
+If :ref:`ReportMoreUnsafeFunctions<option-ReportMoreUnsafeFunctions>` is enabled,
```
```suggestion
If :option:`ReportMoreUnsafeFunctions` is enabled,
```

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
+Custom functions
+----------------
+
+The option `CustomFunctions` allows the user to define custom functions to be
```
```suggestion
The option :option:`CustomFunctions` allows the user to define custom functions to be
```

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
+   "
+
+The functions are matched using POSIX extended regular expressions.
+*(Note: The regular expressions do not support negative* ``(?!)`` *matches)*
```
```suggestion
*(Note: The regular expressions do not support negative* ``(?!)`` *matches.)*
```

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -74,10 +75,50 @@ Both macros have to be defined to suggest replacement functions from *Annex K.*
 Options
 -------
 
+.. _option-ReportMoreUnsafeFunctions:
+
```
```suggestion
```

With using `:option:...` maybe these are not needed, but please double check!

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -86,6 +127,21 @@ Options
    this option enables.
    Default is `true`.
 
+.. _option-ReportDefaultFunctions:
+
```
```suggestion
```

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/unsafe-functions.rst:
```diff
@@ -86,6 +127,21 @@ Options
+.. option:: ReportDefaultFunctions
+
+    When `true`, the check reports the default set of functions.
+    Copnsider changing the setting to false if you only want to see custom
```
```suggestion
    Consider changing the setting to false if you only want to see custom
```

Typo!

Discookie committed:
Merge branch 'main' into unsafe-expand

Discookie committed:
Fix docs againagain

Discookie committed:
Reword release notes

Discookie committed:
Merge branch 'main' into unsafe-expand

Discookie commented:
Fixed the outstanding minor issues.
